### ISSUE TITLE
Cannot read property 'checkboxes' of undefined

### ISSUE THREAD

**Issue by s0By** [gyrocode/jquery-datatables-checkboxes#23]
With the updated code I think I found a bug. Previously with the older code it worked. 
I have an external button that "clicks" the select all button, something like this:
```
$('.bulk-actions .btn--selection').on('click', function(e){
     $('.dt-checkboxes-select-all').trigger('click');
});
```
The click event itself works, I see the checkboxes change and the 
`$(table.table().container()).on('change', 'input[type="checkbox"]', function() {` grabs the count, but it throws and error as well:

![screen shot 2017-02-01 at 14 44 30](https://cloud.githubusercontent.com/assets/6382468/22507285/91ab281e-e88d-11e6-80ec-b41e2ea02095.png)

Lines may not correspond, because I'm playing around with custom checkboxes from bootstrap, mentioned that a while back to you, but you're still having doubts. Maybe I'll make a pull request later.

The lines in my case are:

275: `self.onClickSelectAll(e, this);`
545: `(ctx.aoColumns[col].checkboxes && ctx.aoColumns[col].checkboxes.selectAllPages)`

**Comment by mpryvkin** [comment]
I recommend to use API, for example [`column().checkboxes.select()`](http://www.gyrocode.com/projects/jquery-datatables-checkboxes/reference/api/#column().checkboxes.select()) to select all checkboxes in the column rather than use `trigger()`.

Internally I use delegated event handlers or cancellation of event propagation for various reasons.  Most likely that's what's causing unexpected behavior with `trigger()`.

**Comment by s0By** [comment]
Yes, one way to do it. 

`trigger('click')` was more comfortable because I didn't have to determine wether to check or uncheck and the `on change` listener worked.

So here's another approach which works in current situation: 
```
var select_all = $(table.table().container()).find('.dt-checkboxes-select-all');
select_all.prop('indeterminate') ? table.column(0).checkboxes.select() : table.column(0).checkboxes.deselect();
select_all.trigger('change');
```

Perhaps we could reopen the fixedColumns issue, that would be nice to fix. I'll take a look into it as well.

**Comment by mpryvkin** [comment]
Why are you doing all of this? "Select all" control in the header does it automatically. Do you have "select all" control outside of the table? 

Also why do you mention FixedColumns?

**Comment by mpryvkin** [comment]
Latest version of the code works correctly with FixedColumns extension, there is no need for manually handling "Select all" control that has been duplicated by FixedColumns extension.

**Comment by s0By** [comment]
In my use case I have a table of items:
![screen shot 2017-03-24 at 15 56 33](https://cloud.githubusercontent.com/assets/6382468/24297245/bea93370-10aa-11e7-9f30-0ab3f3f125a6.png)

When I click on "select-all" or some item from the table I get extra controls over the items:

![screen shot 2017-03-24 at 15 57 02](https://cloud.githubusercontent.com/assets/6382468/24297246/bea932c6-10aa-11e7-9852-fdeaeb16a799.png)

"1 product selected" is like a "mask" button above the checkbox and should work like a button and a checkbox is a part of it, but since it's external, I had to implement the solution mentioned above.

---
Mentioned fixedColumns's [issue](https://github.com/gyrocode/jquery-datatables-checkboxes/issues/13) because previously it had a bunch of issues and it was really hard to communicate, your answers would only get to me via email, I don't see them inside the issue after it got closed. I'll check the latest version if it's good to go.

**Comment by mpryvkin** [comment]
That explains it, please let me know what you find out.

Just curious, what plugin do you use to style your checkboxes?

**Comment by mpryvkin** [comment]
It looks like [awesome-bootstrap-checkbox](http://flatlogic.github.io/awesome-bootstrap-checkbox).

**Comment by hycday** [comment]
I don't get it, how do you get the "1 product selected" / "pin product" / "clear" and "edit products" options ?! I like that !
for the style, I would say some bootstrap css

**Comment by s0By** [comment]
My own solution, have some different looks, still playing around with it. Had to rewrite a part of your code though, mentioned that to you sometime ago. I'm thinking of a robust/dynamic solution then I'll make a pull request. It's a new approach which is widely used now to get custom look of radio/checkbox buttons, you have to wrap input inside label 
```
<label class="custom-control custom-checkbox">
<input type="checkbox" class="custom-control-input dt-checkboxes">
<span class="custom-control-indicator"></span>
</label>
```
Span tag is needed for the text (if you have any). awesome-bootstrap-checkbox is a more older solution, I used similar approach sometime ago, but it's harder to achieve the same look in every browser since the checkbox itself is actually displayed using ::before ::after

**Comment by s0By** [comment]
It seems that it works, hard to tell at the moment. Global change listener is still not working, but I don't think that's your issue, when fixedColumn is active it just stops working for some reason:
```
$(table.table().container()).on('change', 'input[type="checkbox"]', function() {
}
```
Even rowCallback doesn't function properly, row click doesn't work:
```
rowCallback: function (row, data, index) {
      var _this = this;

     $(row).click(function (e) {
             var target = $(e.target);
             console.log(target);
      }
});
```
Did you have any experience with this case?

**Comment by mpryvkin** [comment]
I don't like FixedColumns extension, it causes appearance issues in my tests. I would avoid using it on any production system, would rather use Responsive extension to display tables with columns that span beyond window width.

I managed to add support for it for checkboxes but I can't tell you much for your scenario.

**Comment by s0By** [comment]
I agree, it's not the best extension. I will try to contact allan and see what he says. I wanted to stick to css solution in the beginning, but it was really hard to apply on datatable, scrollY stopped working with it etc. Can't have mile long tables on pages.

For my system it's hard to avoid it because Responsive extension just doesn't cut it, clients have the ability to select what data they want to see, so basically when you select 20 columns, you need to somehow see them and not hide them. 

By the way, if anyone will be interested in external click button for the click all, had to tweak it a bit, since you need to consider "not clicked" when having a server-side solution: 
```
var select_all = $(table.table().container()).find('.dt-checkboxes-select-all input');
                    if(select_all.prop('indeterminate') || !select_all.prop('checked')) {
                        table.column(0).checkboxes.select();
                    } else {
                        table.column(0).checkboxes.deselect();
                    }

                    select_all.trigger('change');
 });
```

**Comment by s0By** [comment]
Houston, we have a problem. 
When started to write and prepare a test case about the fixedColumn issue for Allan I noticed that actually everything works, here's [jsFiddle ](https://jsfiddle.net/ilyakapitan/d5Ld0x7n/8/).

So I couldn't believe myself, started to debug my solution with no luck. Then I turned off the external checkbox.js and pasted inside the code of 1.2.0 and it stopped working...

Something changed in code between 1.0.3 and 1.2.0 that is blocking the change event when combined with fixedColumns

P.S. pasted 1.1.0 it works there as well. Could you take a look what could be the difference? Trying to find it as well.

**Comment by mpryvkin** [comment]
Generally I don't recommend listening for 'change' event for the reason you've discovered. In 1.2.0 I added better support for FixedColumn where checkbox could be checked via API and I update cloned checkbox in `updateFixedColumn()` function. Most likely this negates your click handler since elements in the cloned column get recreated.

In 1.2.0 I added callbacks for select/deselect events that could be defined using options `selectCallback` and `selectAllCallback`. They are not yet documented.

See [this example](https://jsfiddle.net/d5Ld0x7n/9/) for correct way now to handle select/deselect event.
